onnxruntime/core/providers/cuda/math/matmul.cc

```diff
@@ -88,6 +95,366 @@ static bool CanUseStridedBatchedGemm(const TensorShape& left_shape, const Tensor
+  float squaredSum = 0.0f;
+  for (int32_t i = 0; i < size; i ++) {
+    float diff = float(elems[i]) - mean;
+    squaredSum += diff * diff;
+  }
+
+  // Calculate the variance
+  float variance = squaredSum / size;
+
+  // Return the square root of variance as standard deviation
+  return std::sqrt(variance);
+}
+
+Status ComputeScale(cudaStream_t& stream, const Tensor* tensor, const float std_quant, float& scale)
+{
+  const int32_t num_coef = tensor->Shape().Size();
+  MLFloat16* scale_coef = (MLFloat16*)malloc(num_coef * sizeof(MLFloat16));
+  auto status = ComputeStdDevCoefficientsForScale(stream, tensor, num_coef, scale_coef);
+  if (! status.IsOK())
+    return status;
+
+  float std_coef = ComputeStandardDeviation(scale_coef, num_coef);
+  free(scale_coef);
+
+  // If the standard deviation is 0, just use a scale of 1
+  scale = fabs(std_coef) < 1e-5 ? 1.0f : std_quant / std_coef;
+
+  return status;
+}
+
+void NoOpDeleter(void* [[maybe_unused]] ptr) {
+  (void)(ptr);
+}
+
+Status TransposeAndConvertToFp8(cudaStream_t& cuda_stream, const Tensor* right_X, IAllocatorUniquePtr<void>& right_X_fp8,
+  const cudaDeviceProp& device_prop, const AllocatorPtr& allocator, Stream* stream)
+{
+  const auto shape_size = right_X->Shape().GetDims().size();
+  const int num_elems = right_X->SizeInBytes() / sizeof(MLFloat16);
+
+  cublasHandle_t cublas_handle;
+  CUBLAS_RETURN_IF_ERROR(cublasCreate(&cublas_handle));
+
+  // We don't have access to the context from inside PrePack, which is one of the callers of this method.
+  right_X_fp8 = IAllocator::MakeUniquePtr<void>(allocator, num_elems * sizeof(Float8E4M3FN), false, stream);
+
+  Status return_status;
+  if (shape_size > 1)
+  {
+    InlinedVector<size_t> perm;
+    perm.push_back(1);
+    perm.push_back(0);
+    for (size_t i = 2; i < shape_size; i++) {
+      perm.push_back(i);
+    }
+    gsl::span<size_t> permutation(perm.data(), shape_size);
+
+    TensorShapeVector dims;
+    for (size_t i = 0; i < shape_size; i++) {
+      dims.push_back(right_X->Shape()[perm[i]]);
+    }
+
+    std::unique_ptr<Tensor> right_X_transposed = Tensor::Create(right_X->DataType(), TensorShape(dims), allocator);
+
+    auto status = cuda::Transpose::DoTranspose(device_prop, cuda_stream, cublas_handle, permutation, *right_X, *right_X_transposed);
+    if (!status.IsOK()) {
+      return status;
+    }
+    return_status = MLFloat16ToFloat8E4M3FN(cuda_stream, right_X_transposed.get(), right_X_fp8.get());
+  }
+  else
+  {
+    return_status = MLFloat16ToFloat8E4M3FN(cuda_stream, right_X, right_X_fp8.get());
+  }
+
+  CUBLAS_RETURN_IF_ERROR(cublasDestroy(cublas_handle));
+  return return_status;
+}
+
+Status ComputeUsingFp8(OpKernelContext* ctx, MatMulComputeHelper& helper,  cudaStream_t& stream,
+  const cudaDeviceProp& device_prop, const cublasLtEpilogue_t& epilogue, float alpha,
+  IAllocatorUniquePtr<void>& right_X_fp8)
+{
+  CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(stream));
+
+  cublasLtHandle_t cublasLt;
```

Comment:
You can use this to get a handle: https://github.com/microsoft/onnxruntime/blob/deee48002cf63edc078c4a50ac6e9ac0b069ae85/onnxruntime/core/providers/cuda/cuda_kernel.h#L175

Reply:
updated